fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java

```diff
@@ -0,0 +1,609 @@
+
+        if (info.outputStringColumns.isEmpty()) {
+            return DecodeNodeInfo.EMPTY;
+        }
+
+        return info;
+    }
+
+    private void collectPredicate(Operator operator, DecodeNodeInfo info) {
+        if (operator.getPredicate() == null) {
+            return;
+        }
+        DictExpressionCollector dictExpressionCollector = new DictExpressionCollector(info.outputStringColumns);
+        dictExpressionCollector.collect(operator.getPredicate());
+
+        info.outputStringColumns.getStream().forEach(c -> {
+            List<ScalarOperator> expressions = dictExpressionCollector.getDictExpressions(c);
+            if (!expressions.isEmpty()) {
+                // predicate only translate to string expression
+                stringExpressions.computeIfAbsent(c, l -> Lists.newArrayList()).addAll(expressions);
+            }
+        });
+    }
+
+    private void collectProjection(Operator operator, DecodeNodeInfo info) {
+        if (operator.getProjection() == null) {
+            return;
+        }
+
+        ColumnRefSet decodeInput = info.outputStringColumns;
+        info.outputStringColumns = new ColumnRefSet();
+        for (ColumnRefOperator key : operator.getProjection().getColumnRefMap().keySet()) {
+            if (decodeInput.contains(key)) {
+                info.outputStringColumns.union(key.getId());
+                continue;
+            }
+
+            DictExpressionCollector dictExpressionCollector = new DictExpressionCollector(decodeInput);
+
+            ScalarOperator value = operator.getProjection().getColumnRefMap().get(key);
+            dictExpressionCollector.collect(value);
+
+            decodeInput.getStream().forEach(c -> {
+                // collect dict expression
+                List<ScalarOperator> exprs = dictExpressionCollector.getDictExpressions(c);
+                if (!exprs.isEmpty()) {
+                    // maybe not new dict, just optimize the expression with dictionary
+                    stringExpressions.computeIfAbsent(c, l -> Lists.newArrayList()).addAll(exprs);
+                }
+
+                // whole expression support dictionary, define new dict column, only support varchar column
+                if (exprs.contains(value) && value.getType().isVarchar()) {
+                    stringRefToDefineExprMap.put(key.getId(), value);
+                    expressionStringRefCounter.putIfAbsent(key.getId(), 0);
+                    info.outputStringColumns.union(key.getId());
+                }
+            });
+        }
+    }
+
+    // Check if an expression can be optimized using a dictionary
+    // If the expression only contains a string column, the expression can be optimized using a dictionary
+    private static class DictExpressionCollector extends ScalarOperatorVisitor<ScalarOperator, Void> {
+        private static final ScalarOperator CONSTANTS = ConstantOperator.TRUE;
+        private static final ScalarOperator VARIABLES = ConstantOperator.FALSE;
```

Comment:
VARIABLES is a little confued? how about DICT_COL_HODLER ?

Reply:
I want to mark the child not only one string column, so DICT_COL_HODLER is not good name? or other better name?

Reply:
VARIABLES means expression contains other-variable-column or we think we should take the expression has other-variable-column. How do we deal with return VARIABLE's dictionary pain?

Reply:
keep origin expression and don't handle it if return `VARIABLES`,